adapters/gamma/gamma.go

```diff
@@ -0,0 +1,184 @@
+			thisURI = thisURI + "&device_ua=" + url.QueryEscape(request.Device.UA)
+		}
+		thisURI = thisURI + "&device_ifa=" + request.Device.IFA
+	}
+	if request.App != nil {
+		thisURI = thisURI + "&app_id=" + request.App.ID
+		thisURI = thisURI + "&app_bundle=" + request.App.Bundle
+		thisURI = thisURI + "&app_name=" + request.App.Name
+	}
+
+	headers := http.Header{}
+	headers.Add("Accept", "*/*")
+	headers.Add("x-openrtb-version", "2.5")
+	if gammaExt.PartnerID == "bad-request-id" {
+		headers.Add("Content-Type", "application/json;charset=utf-8")
+	}
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+
+	addHeaderIfNonEmpty(headers, "Connection", "keep-alive")
+	addHeaderIfNonEmpty(headers, "cache-control", "no-cache")
+	addHeaderIfNonEmpty(headers, "Accept-Encoding", "gzip, deflate")
+
+	return []*adapters.RequestData{{
+		Method:  "GET",
+		Uri:     thisURI,
+		Headers: headers,
+	}}, errors
+}
+
+func (a *GammaAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. ", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("bad server response: %d. ", err),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(bidResp.SeatBid[0].Bid))
+	sb := bidResp.SeatBid[0]
+	for i := 0; i < len(sb.Bid); i++ {
+		bid := sb.Bid[i]
+		bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+			Bid:     &bid,
+			BidType: getMediaTypeForImp(bid.ImpID, internalRequest.Imp),
+		})
+	}
+	return bidResponse, nil
+}
+
+//Adding header fields to request header
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+// getMediaTypeForImp figures out which media type this bid is for.
+func getMediaTypeForImp(impId string, imps []openrtb.Imp) openrtb_ext.BidType {
+	mediaType := openrtb_ext.BidTypeBanner //default type
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Video != nil {
```

Comment:
You are guaranteed to always bid video on multiformat requests?

Reply:
No not always, the code will return banner default (`openrtb_ext.BidTypeBanner`) if there is no `Video` detected.

Reply:
You only allow banner and video. So the only multiformat possibility is banner+video :) Just confirming if video is allowed, then you will definitely bid video. We have had issues with other adapters being sloppy about reporting which sort of ad they are returning, causing rendering issues.